java/libphonenumber/src/com/google/i18n/phonenumbers/SingleFileMetadataSourceImpl.java

```diff
@@ -0,0 +1,98 @@
+ * during initialization.
+ */
+public final class SingleFileMetadataSourceImpl implements MetadataSource {
+
+  private static final Logger logger =
+      Logger.getLogger(SingleFileMetadataSourceImpl.class.getName());
+
+  private static final String META_DATA_FILE =
+      "/com/google/i18n/phonenumbers/data/PhoneNumberMetadataProtoFile";
+
+  // A mapping from a region code to the PhoneMetadata for that region.
+  // Note: Synchronization, though only needed for the Android version of the library, is used in
+  // all versions for consistency.
+  private final Map<String, PhoneMetadata> regionToMetadataMap =
+      Collections.synchronizedMap(new HashMap<String, PhoneMetadata>());
+
+  // A mapping from a country calling code for a non-geographical entity to the PhoneMetadata for
+  // that country calling code. Examples of the country calling codes include 800 (International
+  // Toll Free Service) and 808 (International Shared Cost Service).
+  // Note: Synchronization, though only needed for the Android version of the library, is used in
+  // all versions for consistency.
+  private final Map<Integer, PhoneMetadata> countryCodeToNonGeographicalMetadataMap =
+      Collections.synchronizedMap(new HashMap<Integer, PhoneMetadata>());
+
+  // It is assumed that metadataLoader is not null.
+  public SingleFileMetadataSourceImpl(MetadataLoader metadataLoader) {
+    InputStream input = metadataLoader.loadMetadata(META_DATA_FILE);
+    if (input == null) {
+      throw new IllegalStateException(
+          "no metadata available for PhoneNumberUtil: " + META_DATA_FILE);
+    }
+    PhoneMetadataCollection metadataCollection = loadMetadataAndCloseInput(input);
+    for (PhoneMetadata metadata : metadataCollection.getMetadataList()) {
+      String regionCode = metadata.getId();
+      int countryCallingCode = metadata.getCountryCode();
+      if (PhoneNumberUtil.REGION_CODE_FOR_NON_GEO_ENTITY.equals(regionCode)) {
+        countryCodeToNonGeographicalMetadataMap.put(countryCallingCode, metadata);
+      } else {
+        regionToMetadataMap.put(regionCode, metadata);
+      }
+    }
+  }
+
+  @Override
+  public PhoneMetadata getMetadataForRegion(String regionCode) {
+    return regionToMetadataMap.get(regionCode);
+  }
+
+  @Override
+  public PhoneMetadata getMetadataForNonGeographicalRegion(int countryCallingCode) {
+    return countryCodeToNonGeographicalMetadataMap.get(countryCallingCode);
+  }
+
+  /**
+   * Loads the metadata protocol buffer from the given stream and closes the stream afterwards. Any
+   * exceptions that occur while reading the stream are propagated (though exceptions that occur
+   * when the stream is closed will be ignored).
+   *
+   * @param source  the non-null stream from which metadata is to be read.
+   * @return        the loaded metadata protocol buffer.
+   */
+  private static PhoneMetadataCollection loadMetadataAndCloseInput(InputStream source) {
```

Comment:
Slightly inconsistent with MultiFileMetadataSourceImpl.java here: there, loadMetadataAndCloseInput takes in an ObjectInputStream already. It doesn't matter too much since these are private, but it'd be nice if they were consistent.